cups/man/lp.man
.\"
.\" "$Id: lp.man,v 1.3 2005/01/04 22:10:44 jlovell Exp $"
.\"
.\"   lp/cancel man page for the Common UNIX Printing System (CUPS).
.\"
.\"   Copyright 1997-2005 by Easy Software Products.
.\"
.\"   These coded instructions, statements, and computer programs are the
.\"   property of Easy Software Products and are protected by Federal
.\"   copyright law.  Distribution and use rights are outlined in the file
.\"   "LICENSE.txt" which should have been included with this file.  If this
.\"   file is missing or damaged please contact Easy Software Products
.\"   at:
.\"
.\"       Attn: CUPS Licensing Information
.\"       Easy Software Products
.\"       44141 Airport View Drive, Suite 204
.\"       Hollywood, Maryland 20636 USA
.\"
.\"       Voice: (301) 373-9600
.\"       EMail: cups-info@cups.org
.\"         WWW: http://www.cups.org
.\"
.TH lp 1 "Common UNIX Printing System" "1 May 2003" "Easy Software Products"
.SH NAME
lp \- print files
.br
cancel \- cancel jobs
.SH SYNOPSIS
.B lp
[ -E ] [ \-c ] [ \-d
.I destination
] [ \-h
.I server
] [ \-m ] [ \-n
.I num-copies
[ \-o
.I option
] [ \-q
.I priority
] [ \-s ] [ \-w
] [ \-t
.I title
] [ \-H
.I handling
] [ \-P
.I page-list
] [
.I file(s)
]
.br
.B lp
[ -E ] [ \-c ] [ \-h
.I server
] [ \-i
.I job-id
] [ \-n
.I num-copies
[ \-o
.I option
] [ \-q
.I priority
] [ \-t
.I title
] [ \-H
.I handling
] [ \-P
.I page-list
]
.br
.B cancel
[ \-a ] [ -h
.I server
] [ -u 
.I username
] [
.I id
] [
.I destination
] [
.I destination-id
]
.SH DESCRIPTION
\fBlp\fR submits files for printing or alters a pending job. Use
a filename of "-" to force printing from the standard input.
.LP
\fBcancel\fR cancels existing print jobs. The \fI-a\fR option will remove
all jobs from the specified destination.
.SH OPTIONS
The following options are recognized by \fBlp\fR:
.TP 5
\-E
.br
Forces encryption when connecting to the server.
.TP 5
\-c
.br
This option is provided for backwards-compatibility only. On
systems that support it, this option forces the print file to be
copied to the spool directory before printing. In CUPS, print
files are always sent to the scheduler via IPP which has the
same effect.
.TP 5
\-d \fIdestination\fR
.br
Prints files to the named printer.
.TP 5
\-h \fIhostname\fR
.br
Specifies the print server hostname. The default is "localhost" or the value
of the CUPS_SERVER environment variable.
.TP 5
\-i \fIjob-id\fR
.br
Specifies an existing job to modify.
.TP 5
\-m
.br
Send email when the job is completed (not supported CUPS 1.1.)
.TP 5
\-n \fIcopies\fR
.br
Sets the number of copies to print from 1 to 100.
.TP 5
\-o \fIoption\fR
.br
Sets a job option.
.TP 5
\-q \fIpriority\fR
.br
Sets the job priority from 1 (lowest) to 100 (highest). The
default priority is 50.
.TP 5
\-s
.br
Do not report the resulting job IDs (silent mode.)
.TP 5
\-t \fIname\fR
.br
Sets the job name.
.TP 5
\-u \fIusername\fR
.br
Cancels jobs owned by \fIusername\fR.
.TP 5
\-w
.br
Report the resulting job IDs.
.TP 5
\-H \fIhandling\fR
.br
Specifies when the job should be printed. A value of
\fIimmediate\fR will print the file immediately, a value of
\fIhold\fR will hold the job indefinitely, and a time value
(HH:MM) will hold the job until the specified time. Use a value
of \fIresume\fR with the \fI-i\fR option to resume a held job.
Use a value of \fIrestart\fR with the \fI-i\fR option to restart
a completed job.
.TP 5
\-P \fIpage-list\fR
.br
Specifies which pages to print in the document. The list can contain a
list of numbers and ranges (#-#) separated by commas (e.g. 1,3-5,16).
.LP
The interpretation of options may be cancelled by the option `--'. This
will allow the use of file names that begin with a dash (`-'). For example:
.D1
lp -- -filename
.SH COMPATIBILITY
Unlike the System V printing system, CUPS allows printer names to contain
any printable character except SPACE and TAB. Also, printer and class names are
\fBnot\fR case-sensitive. 
.LP
The "m" option is not functional in CUPS 1.1.
.LP
The "q" option accepts a different range of values than the
Solaris lp command, matching the IPP job priority values (1-100)
instead of the Solaris values (0-39).
.SH NOTES
Administrators wishing to prevent unauthorized cancellation of
jobs via the \fI-u\fR option should password-protect the /jobs
location in \fIcupsd.conf(5)\fR.
.SH SEE ALSO
lpstat(1),
CUPS Software Users Manual,
http://localhost:631/documentation.html
.SH COPYRIGHT
Copyright 1993-2005 by Easy Software Products, All Rights Reserved.
.\"
.\" End of "$Id: lp.man,v 1.3 2005/01/04 22:10:44 jlovell Exp $".
.\"


X11/ChangeLog

== 10.4 ChangeLog items ==

2005-03-09  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)

2004-10-25  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/InfoPlist.strings:
	removed CFBundleShortVersionString and CFBundleGetInfoString
	(3851089)

2004-10-19  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	added "Terminal" to the strings file	

	* xc/programs/Xserver/hw/apple/X11Application.m (prefs_get):
	added code to call NSLocalizedString on each name in the
	default applications menu

	* xc/programs/Xserver/hw/apple/rootless-window.c
	(RootlessKnowsWindowNumber): added check for zero window id

2004-10-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	Expanded backslash escape sequences since .strings files don't
	seem to handle them correctly. (3833239)

2004-10-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/vfb/InitOutput.c: workaround 3834723
	by rearranging include order

2004-10-10  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Controller.m: QuartzRunClient
	function now has a second arg - whether or not to quote the
	command (3816789, 3572554)

2004-10-08  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/parse.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/Attrib.c: merged Xpm changes from XFree86
	HEAD branch (3744542, 3778302)

2004-09-29  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: fixed
	typo (3820725)

2004-09-16  John Harper  <jsh@apple.com>

	* xc/programs/xterm/main.c: call openpty() to scan for a pseudo
	terminal (3768572)

	* xc/programs/xterm/xterm.man: fix documentation of -ls option
	(3803120)

2004-09-08  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: nib
	changes for 3313133 and 3316468

2004-09-03  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: gave
	applications dialog window a minimum size

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	new file - ran genstrings on the .m files

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	added Localizable.strings

	* xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: added comments
	for translators

	* xc/programs/Xserver/hw/apple/Imakefile: set OBJROOT and
	SYMROOT when calling xcodebuild

2004-08-25  John Harper  <jsh@apple.com>

	* quartz-wm/x-window.m: don't grab Mod1-grave for window
	cycling, only Mod2 (3774967)

2004-08-23  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix typo in
	$XINITRC code (3621248)

2004-07-16  John Harper  <jsh@apple.com>

	* merged changes from XFree86 4.4.0 and fixed resulting
	brokenness (3570235, 3612446, 3548675, 3404805)

2004-06-10  John Harper  <jsh@apple.com>

	* xinitrc: quote shell expanded filenames to support home
	directories with spaces (3674992)

2004-03-29  John Harper  <jsh@apple.com>

	* FS/FSFontInfo.c, FS/FSFtNames.c, FS/FSGetCats.c,
	FS/FSListCats.c, FS/FSListExt.c, FS/FSOpenServ.c,
	FS/FSQGlyphs.c, FS/FSQXExt.c, FS/FSQXInfo.c, FS/FSlibint.h,
	FS/FSlibos.h, font/fc/fsconvert.c, font/fc/fserve.c,
	font/fc/fslibos.h, font/fontfile/dirfile.c,
	font/fontfile/encparse.c, font/fontfile/fontdir.c,
	font/fontfile/fontfile.c: merged font security fixes from
	XFree86 4.3 branch

2004-03-11  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzRunClient):
	added command quoting hack to fix 3572554

	* xc/programs/Xserver/hw/apple/bundle-main.c (define_named):

cups/systemv/lp.c
/*
 * "$Id: lp.c,v 1.7 2005/01/04 22:10:52 jlovell Exp $"
 *
 *   "lp" command for the Common UNIX Printing System (CUPS).
 *
 *   Copyright 1997-2005 by Easy Software Products.
 *
 *   These coded instructions, statements, and computer programs are the
 *   property of Easy Software Products and are protected by Federal
 *   copyright law.  Distribution and use rights are outlined in the file
 *   "LICENSE.txt" which should have been included with this file.  If this
 *   file is missing or damaged please contact Easy Software Products
 *   at:
 *
 *       Attn: CUPS Licensing Information
 *       Easy Software Products
 *       44141 Airport View Drive, Suite 204
 *       Hollywood, Maryland 20636 USA
 *
 *       Voice: (301) 373-9600
 *       EMail: cups-info@cups.org
 *         WWW: http://www.cups.org
 *
 * Contents:
 *
 *   main()          - Parse options and send files for printing.
 *   restart_job()   - Restart a job.
 *   set_job_attrs() - Set job attributes.
 *   sighandler()    - Signal catcher for when we print from stdin...
 */

/*
 * Include necessary headers...
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <cups/cups.h>
#include <cups/string.h>
#include <cups/language.h>


#ifndef WIN32
#  include <unistd.h>
#  include <signal.h>


/*
 * Local functions.
 */

void	sighandler(int);
#endif /* !WIN32 */
int	restart_job(int job_id);
int	set_job_attrs(int job_id, int num_options, cups_option_t *options);


/*
 * Globals...
 */

char	tempfile[1024];		/* Temporary file for printing from stdin */


/*
 * 'main()' - Parse options and send files for printing.
 */

int
main(int  argc,		/* I - Number of command-line arguments */
     char *argv[])	/* I - Command-line arguments */
{
  int		i, j;		/* Looping vars */
  int		job_id;		/* Job ID */
  char		*printer,	/* Printer name */
		*instance,	/* Instance name */ 
		*val,		/* Option value */
		*title;		/* Job title */
  int		priority;	/* Job priority (1-100) */
  int		num_copies;	/* Number of copies per file */
  int		num_files;	/* Number of files to print */
  const char	*files[1000];	/* Files to print */
  int		num_dests;	/* Number of destinations */
  cups_dest_t	*dests,		/* Destinations */
		*dest;		/* Selected destination */
  int		num_options;	/* Number of options */
  cups_option_t	*options;	/* Options */
  int		silent,		/* Silent or verbose output? */
		opt_end;	/* End of options */
  char		buffer[8192];	/* Copy buffer */
  int		temp;		/* Temporary file descriptor */
  off_t		size;		/* Temporary file size */
#if defined(HAVE_SIGACTION) && !defined(HAVE_SIGSET)
  struct sigaction action;	/* Signal action */
  struct sigaction oldaction;	/* Old signal action */
#endif /* HAVE_SIGACTION && !HAVE_SIGSET*/


#ifdef __sun
 /*
  * Solaris does some rather strange things to re-queue remote print
  * jobs.  On bootup, the "lp" command is run as "printd" to re-spool
  * any remote jobs in /var/spool/print.  Since CUPS doesn't need this
  * nonsense, we just need to add the necessary check here to prevent
  * lp from causing boot problems...
  */

  if ((val = strrchr(argv[0], '/')) != NULL)
    val ++;
  else
    val = argv[0];

  if (strcmp(val, "printd") == 0)
    return (0);
#endif /* __sun */

  opt_end     = 0;
  silent      = 0;
  printer     = NULL;
  num_dests   = 0;
  dests       = NULL;
  num_options = 0;
  options     = NULL;
  num_files   = 0;
  title       = NULL;
  job_id      = 0;

  for (i = 1; i < argc; i ++)
    if (argv[i][0] == '-' && argv[i][1] && !opt_end)
      switch (argv[i][1])
      {
        case 'E' : /* Encrypt */
#ifdef HAVE_SSL
	    cupsSetEncryption(HTTP_ENCRYPT_REQUIRED);
#else
            fprintf(stderr, "%s: Sorry, no encryption support compiled in!\n",
	            argv[0]);
#endif /* HAVE_SSL */
	    break;

        case 'c' : /* Copy to spool dir (always enabled) */
	    break;

        case 'd' : /* Destination printer or class */
	    if (argv[i][2] != '\0')
	      printer = argv[i] + 2;
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected destination after -d option!\n", stderr);
		return (1);
              }

	      printer = argv[i];
	    }

            if ((instance = strrchr(printer, '/')) != NULL)
	      *instance++ = '\0';

	    if (num_dests == 0)
	      num_dests = cupsGetDests(&dests);

            if ((dest = cupsGetDest(printer, instance, num_dests, dests)) != NULL)
	    {
	      for (j = 0; j < dest->num_options; j ++)
	        if (cupsGetOption(dest->options[j].name, num_options, options) == NULL)
	          num_options = cupsAddOption(dest->options[j].name,
		                              dest->options[j].value,
					      num_options, &options);
	    }
	    break;

        case 'f' : /* Form */
	    if (!argv[i][2])
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected form after -f option!\n", stderr);
		return (1);
              }
	    }

	    fputs("lp: Warning - form option ignored!\n", stderr);
	    break;

        case 'h' : /* Destination host */
	    if (argv[i][2] != '\0')
	      cupsSetServer(argv[i] + 2);
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected hostname after -h option!\n", stderr);
		return (1);
              }

	      cupsSetServer(argv[i]);
	    }
	    break;

        case 'i' : /* Change job */
	    if (argv[i][2])
	      val = argv[i] + 2;
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected job ID after -i option!\n", stderr);
		return (1);
              }

	      val = argv[i];
	    }

            if (num_files > 0)
	    {
	      fputs("lp: Error - cannot print files and alter jobs simultaneously!\n", stderr);
	      return (1);
	    }

            if (strrchr(val, '-') != NULL)
	      job_id = atoi(strrchr(val, '-') + 1);
	    else
	      job_id = atoi(val);

            if (job_id < 0)
	    {
	      fputs("lp: Error - bad job ID!\n", stderr);
	      break;
	    }
	    break;

	case 'm' : /* Send email when job is done */
#ifdef __sun
	case 'p' : /* Notify on completion */
#endif /* __sun */
	    break;

	case 'w' : /* Write to console or email */
	    silent = 0;
	    break;

	case 'n' : /* Number of copies */
	    if (argv[i][2] != '\0')
	      num_copies = atoi(argv[i] + 2);
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected copies after -n option!\n", stderr);
		return (1);
              }

	      num_copies = atoi(argv[i]);
	    }

            sprintf(buffer, "%d", num_copies);
            num_options = cupsAddOption("copies", buffer, num_options, &options);
	    break;

	case 'o' : /* Option */
	    if (argv[i][2] != '\0')
	      num_options = cupsParseOptions(argv[i] + 2, num_options, &options);
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected option string after -o option!\n", stderr);
		return (1);
              }

	      num_options = cupsParseOptions(argv[i], num_options, &options);
	    }
	    break;

#ifndef __sun
	case 'p' : /* Queue priority */
#endif /* !__sun */
	case 'q' : /* Queue priority */
	    if (argv[i][2] != '\0')
	      priority = atoi(argv[i] + 2);
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fprintf(stderr, "lp: Expected priority after -%c option!\n",
		        argv[i - 1][1]);
		return (1);
              }

	      priority = atoi(argv[i]);
	    }

           /*
	    * For 100% Solaris compatibility, need to add:
	    *
	    *   priority = 99 * (39 - priority) / 39 + 1;
	    *
	    * However, to keep CUPS lp the same across all platforms
	    * we will break compatibility this far...
	    */

	    if (priority < 1 || priority > 100)
	    {
	      fputs("lp: Priority must be between 1 and 100.\n", stderr);
	      return (1);
	    }

            sprintf(buffer, "%d", priority);
            num_options = cupsAddOption("job-priority", buffer, num_options, &options);
	    break;

	case 's' : /* Silent */
	    silent = 1;
	    break;

	case 't' : /* Title */
	    if (argv[i][2] != '\0')
	      title = argv[i] + 2;
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected title after -t option!\n", stderr);
		return (1);
              }

	      title = argv[i];
	    }
	    break;

        case 'y' : /* mode-list */
	    if (!argv[i][2])
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected mode list after -y option!\n", stderr);
		return (1);
              }
	    }

	    fputs("lp: Warning - mode option ignored!\n", stderr);
	    break;

        case 'H' : /* Hold job */
	    if (argv[i][2])
	      val = argv[i] + 2;
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected hold name after -H option!\n", stderr);
		return (1);
              }

	      val = argv[i];
	    }

	    if (strcmp(val, "hold") == 0)
              num_options = cupsAddOption("job-hold-until", "indefinite",
	                                  num_options, &options);
	    else if (strcmp(val, "resume") == 0 ||
	             strcmp(val, "release") == 0)
              num_options = cupsAddOption("job-hold-until", "no-hold",
	                                  num_options, &options);
	    else if (strcmp(val, "immediate") == 0)
              num_options = cupsAddOption("job-priority", "100",
	                                  num_options, &options);
	    else if (strcmp(val, "restart") == 0)
	    {
	      if (job_id < 1)
	      {
	        fputs("lp: Need job ID (-i) before \"-H restart\"!\n", stderr);
		return (1);
	      }

	      if (restart_job(job_id))
	        return (1);
	    }
	    else
              num_options = cupsAddOption("job-hold-until", val,
	                                  num_options, &options);
	    break;

        case 'P' : /* Page list */
	    if (argv[i][2])
	      val = argv[i] + 2;
	    else
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected page list after -P option!\n", stderr);
		return (1);
              }

	      val = argv[i];
	    }

            num_options = cupsAddOption("page-ranges", val, num_options,
	                                &options);
            break;

        case 'S' : /* character set */
	    if (!argv[i][2])
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected character set after -S option!\n", stderr);
		return (1);
              }
	    }

	    fputs("lp: Warning - character set option ignored!\n", stderr);
	    break;

        case 'T' : /* Content-Type */
	    if (!argv[i][2])
	    {
	      i ++;

	      if (i >= argc)
	      {
	        fputs("lp: Expected content type after -T option!\n", stderr);
		return (1);
              }
	    }

	    fputs("lp: Warning - content type option ignored!\n", stderr);
	    break;

        case '-' : /* End of options */
	    if (argv[i][2])
	    {
	      fputs("lp: Unexpected character after -- option!\n", stderr);
	      return (1);
	    }

            opt_end = 1;
	    break;

	default :
	    fprintf(stderr, "lp: Unknown option \'%c\'!\n", argv[i][1]);
	    return (1);
      }
    else if (!strcmp(argv[i], "-"))
    {
      if (num_files || job_id)
      {
        fputs("lp: Error - cannot print from stdin if files or a job ID are "
	      "provided!\n", stderr);
	return (1);
      }

      break;
    }
    else if (num_files < 1000 && job_id == 0)
    {
     /*
      * Print a file...
      */

      if (access(argv[i], R_OK) != 0)
      {
        fprintf(stderr, "lp: Unable to access \"%s\" - %s\n", argv[i],
	        strerror(errno));
        return (1);
      }

      files[num_files] = argv[i];
      num_files ++;

      if (title == NULL)
      {
        if ((title = strrchr(argv[i], '/')) != NULL)
	  title ++;
	else
          title = argv[i];
      }
    }
    else
      fprintf(stderr, "lp: Too many files - \"%s\"\n", argv[i]);

 /*
  * See if we are altering an existing job...
  */

  if (job_id)
    return (set_job_attrs(job_id, num_options, options));

 /*
  * See if we have any files to print; if not, print from stdin...
  */

  if (printer == NULL)
  {
    if (num_dests == 0)
      num_dests = cupsGetDests(&dests);

    if ((dest = cupsGetDest(NULL, NULL, num_dests, dests)) != NULL)
    {
      printer = dest->name;

      for (j = 0; j < dest->num_options; j ++)
	if (cupsGetOption(dest->options[j].name, num_options, options) == NULL)
	  num_options = cupsAddOption(dest->options[j].name,
		                      dest->options[j].value,
				      num_options, &options);
    }
  }

  if (printer == NULL)
  {
    val = NULL;

    if ((printer = getenv("LPDEST")) == NULL)
    {
      if ((printer = getenv("PRINTER")) != NULL)
      {
        if (!strcmp(printer, "lp"))
          printer = NULL;
	else
	  val = "PRINTER";
      }
    }
    else
      val = "LPDEST";

    if (printer && !cupsGetDest(printer, NULL, num_dests, dests))
      fprintf(stderr, "lp: error - %s environment variable names non-existent destination \"%s\"!\n",
              val, printer);
    else if (cupsLastError() == IPP_NOT_FOUND)
      fputs("lp: error - no default destination available.\n", stderr);
    else
      fputs("lp: error - scheduler not responding!\n", stderr);

    return (1);
  }

  if (num_files > 0)
    job_id = cupsPrintFiles(printer, num_files, files, title, num_options, options);
  else
  {
    num_files = 1;

#ifndef WIN32
#  if defined(HAVE_SIGSET)
    sigset(SIGHUP, sighandler);
    if (sigset(SIGINT, sighandler) == SIG_IGN)
      sigset(SIGINT, SIG_IGN);
    sigset(SIGTERM, sighandler);
#  elif defined(HAVE_SIGACTION)
    memset(&action, 0, sizeof(action));
    action.sa_handler = sighandler;

    sigaction(SIGHUP, &action, NULL);
    sigaction(SIGINT, NULL, &oldaction);
    if (oldaction.sa_handler != SIG_IGN)
      sigaction(SIGINT, &action, NULL);
    sigaction(SIGTERM, &action, NULL);
#  else
    signal(SIGHUP, sighandler);
    if (signal(SIGINT, sighandler) == SIG_IGN)
      signal(SIGINT, SIG_IGN);
    signal(SIGTERM, sighandler);
#  endif
#endif /* !WIN32 */

    temp = cupsTempFd(tempfile, sizeof(tempfile));

    if (temp < 0)
    {
      fprintf(stderr, "lp: unable to create temporary file \"%s\" - %s\n",
              tempfile, strerror(errno));
      return (1);
    }

    while ((i = read(0, buffer, sizeof(buffer))) > 0)
      write(temp, buffer, i);

    size = lseek(temp, 0, SEEK_CUR);
    close(temp);

    if (size == 0)
    {
      fputs("lp: stdin is empty, so no job has been sent.\n", stderr);
      return (1);
    }

    if (title)
      job_id = cupsPrintFile(printer, tempfile, title, num_options, options);
    else
      job_id = cupsPrintFile(printer, tempfile, "(stdin)", num_options, options);

    unlink(tempfile);
  }

  if (job_id < 1)
  {
    fprintf(stderr, "lp: unable to print file: %s\n",
	    ippErrorString(cupsLastError()));
    return (1);
  }
  else if (!silent)
    printf("request id is %s-%d (%d file(s))\n", printer, job_id, num_files);

  return (0);
}


/*
 * 'restart_job()' - Restart a job.
 */

int					/* O - Exit status */
restart_job(int job_id)			/* I - Job ID */
{
  http_t	*http;			/* HTTP connection to server */
  ipp_t		*request,		/* IPP request */
		*response;		/* IPP response */
  cups_lang_t	*language;		/* Language for request */
  char		uri[HTTP_MAX_URI];	/* URI for job */


  http = httpConnectEncrypt(cupsServer(), ippPort(), cupsEncryption());

  language = cupsLangDefault();

  request = ippNew();
  request->request.op.operation_id = IPP_RESTART_JOB;
  request->request.op.request_id   = 1;

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_CHARSET,
               "attributes-charset", NULL, cupsLangEncoding(language));

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_LANGUAGE,
               "attributes-natural-language", NULL, language->language);

  sprintf(uri, "ipp://localhost/jobs/%d", job_id);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI,
               "job-uri", NULL, uri);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
               "requesting-user-name", NULL, cupsUser());

  if ((response = cupsDoRequest(http, request, "/jobs")) != NULL)
  {
    if (response->request.status.status_code > IPP_OK_CONFLICT)
    {
      fprintf(stderr, "lp: restart-job failed: %s\n",
              ippErrorString(response->request.status.status_code));
      ippDelete(response);
      return (1);
    }

    ippDelete(response);
  }
  else
  {
    fprintf(stderr, "lp: restart-job failed: %s\n",
            ippErrorString(cupsLastError()));
    return (1);
  }

  return (0);
}


/*
 * 'set_job_attrs()' - Set job attributes.
 */

int					/* O - Exit status */
set_job_attrs(int           job_id,	/* I - Job ID */
              int           num_options,/* I - Number of options */
	      cups_option_t *options)	/* I - Options */
{
  http_t	*http;			/* HTTP connection to server */
  ipp_t		*request,		/* IPP request */
		*response;		/* IPP response */
  cups_lang_t	*language;		/* Language for request */
  char		uri[HTTP_MAX_URI];	/* URI for job */


  if (num_options == 0)
    return (0);

  http = httpConnectEncrypt(cupsServer(), ippPort(), cupsEncryption());

  language = cupsLangDefault();

  request = ippNew();
  request->request.op.operation_id = IPP_SET_JOB_ATTRIBUTES;
  request->request.op.request_id   = 1;

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_CHARSET,
               "attributes-charset", NULL, cupsLangEncoding(language));

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_LANGUAGE,
               "attributes-natural-language", NULL, language->language);

  sprintf(uri, "ipp://localhost/jobs/%d", job_id);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI,
               "job-uri", NULL, uri);

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
               "requesting-user-name", NULL, cupsUser());

  cupsEncodeOptions(request, num_options, options);

  if ((response = cupsDoRequest(http, request, "/jobs")) != NULL)
  {
    if (response->request.status.status_code > IPP_OK_CONFLICT)
    {
      fprintf(stderr, "lp: set-job-attributes failed: %s\n",
              ippErrorString(response->request.status.status_code));
      ippDelete(response);
      return (1);
    }

    ippDelete(response);
  }
  else
  {
    fprintf(stderr, "lp: set-job-attributes failed: %s\n",
            ippErrorString(cupsLastError()));
    return (1);
  }

  return (0);
}


#ifndef WIN32
/*
 * 'sighandler()' - Signal catcher for when we print from stdin...
 */

void
sighandler(int s)	/* I - Signal number */
{
 /*
  * Remove the temporary file we're using to print from stdin...
  */

  unlink(tempfile);

 /*
  * Exit...
  */

  exit(s);
}
#endif /* !WIN32 */


/*
 * End of "$Id: lp.c,v 1.7 2005/01/04 22:10:52 jlovell Exp $".
 */
